compiler/env/mem/OMRBackingMemoryAllocator.cpp

```diff
@@ -0,0 +1,125 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+
+#include <stdlib.h>
+#include "codegen/FrontEnd.hpp"                  // for feGetEnv
+
+// Memory logging support
+// Memory allocation can be used very early by the compiler: can rely on printf working
+
+static char *LOGLEVEL = ::feGetEnv("OMRDebug_BackingMemory");
+
+// LOGLEVEL="0"    // no debug output
+// LOGLEVEL="1"    // summary output, about one line per function call
+// LOGLEVEL="2"    // extra detail in functions
+// LOGLEVEL="3"    // highest level of detail in functions
+
+#define MEMLOG(n,s, ...) do { if ((LOGLEVEL != NULL) && ((int)(*LOGLEVEL) >= (int)(n+'0'))) { printf("BckMem %p : " s, this, __VA_ARGS__); } } while (0)
+#define MEMLOG0(n,s)     do { if ((LOGLEVEL != NULL) && ((int)(*LOGLEVEL) >= (int)(n+'0'))) { printf("BckMem %p : " s, this); } } while (0)
+
+
+#include "env/mem/BackingMemoryAllocator.hpp"
+
+
+TR::MemorySegment &
+OMR::BackingMemoryAllocator::allocate(size_t requiredSize, void * hint)
+   {
+   size_t sizeToAllocate = std::max(requiredSize, _minimumAllocationSize);
+   MEMLOG(1, "allocate request %lu will allocate at least %lu\n", requiredSize, sizeToAllocate);
+
+   void *p = _rawAllocator.allocate(sizeToAllocate, hint);
+   TR::MemorySegment *newSegment = new (_rawAllocator) TR::MemorySegment(p, sizeToAllocate);
+   if (newSegment == NULL)
+      throw std::bad_alloc();
+   try
+      {
+      _allocatedSegments.push_front( TR::ref(*newSegment) );
+      }
+   catch (...)
+      {
+      _rawAllocator.deallocate(p, sizeToAllocate);
+      throw;
+      }
+   _bytesAllocated += sizeToAllocate;
+   return *newSegment;
+   }
+
+void
+OMR::BackingMemoryAllocator::deallocate(TR::MemorySegment & segment) throw()
+   {
+   MEMLOG(1, "deallocate %p with size %lu request\n", &segment, segment.size());
+
+   bool found=false;
+   for (auto it = _allocatedSegments.begin(); it != _allocatedSegments.end(); it++)
+      {
+      auto ref = *it;
+      MEMLOG(3, "_allocatedSegments iterating segment %p\n", &(ref.get()));
+      if (segment == ref.get())
+         {
+         _allocatedSegments.erase(it);
+         found=true;
+         break;
+         }
+      }
+   TR_ASSERT(found, "Request to deallocate segment not found in allocated segment list");
+
+   _bytesAllocated -= segment.size();
+
+   // deallocate actual segment memory
+   _rawAllocator.deallocate(segment.base(), segment.size());
+
+   // deallocate the memory that held the TR::MemorySegment
+   _rawAllocator.deallocate(&segment, sizeof(TR::MemorySegment));;
+   }
+
+void
+OMR::BackingMemoryAllocator::deallocateSegments()
+   {
+   while (!_allocatedSegments.empty())
+      {
+      TR::MemorySegment &segment = _allocatedSegments.front().get();
+      deallocate(segment);
```

Comment:
Just curious as to whether `BackingMemoryAllocator` can be extensible, and if so, whether we want to enforce that the `OMR` version of this is guaranteed to get called.

Reply:
I'm not sure that's appropriate: if someone creates an extension of BackingMemoryAllocator that implements a different `allocate()` function then presumably they should have the ability to also provide a different `deallocate()` function that matches?

Someone who did that would probably want to further break down `deallocate()` so they could reuse some parts of it without necessarily using other parts (exercise left for those who need it :) ?)

Reply:
Looking at this a second time, wouldn't this result in the code going through the `_allocatedSegments` list for every iteration of the loop?

Reply:
no, because they start at the same end of the list :). It asks what the head is, then removes the head (so `deallocate` just picks off the first segment). It might be possible to refactor `deallocate` to not rely on it and it probably deserves a comment, but i guess I went the easy way.

For the record, I found the list iteration in the earlier implementation, which I initially duplicated, was not as efficient as it should be, so I tweaked it a bit to make it better. So good catch :) .

Reply:
Heh I see, sneaky :P